Add capability matching rules to GEP-33 (#11828)
gardener/gardener: docs/proposals/33-machine-image-capabilities.md

## Proposal

Introduce a top level capabilities array in the CloudProfile `spec.capabilities`.

```go
type Spec struct {

>     Capabilities []map[string][]string

Reviewer: Is this structure even possible or will we run into Protobuf issues (again)?

Frankly, I don't like this data structure much as it's rather unhandy/confusing for implementations. WDYT about the following proposal instead?

```go
Capabilities []Capability

type Capability struct {
    Name string
    Values []string
}
```

cc @LucaBernstein

Reviewer 2: would also be fine for me.

# CloudProfile Example
spec:
  capabilities:
    - hypervisorType: ["gen2", "gen1"]
    - network: ["accelerated", "standard"]
    - storageAccess: ["NVMe", "SCSI"]
    - secureBoot: ["secure", "none"]
    - bootMode: ["uefi-preferred", "uefi", "legacy-bios"]
    ...
```

Please note the following characteristics:
- The array of capabilities is ordered.

> - The array of each capability is ordered also.

Reviewer 3: ```suggestion
- The array of each capability is also ordered.
```

- The order defines the priority of the values in case of multiple supported images.
- The first value in the array is the most preferred value.
- The last value is the least preferred value.

Example: A machine supports hypervisor `gen2` AND `gen1` and an image version offers `gen1` OR `gen2`.
Then the image with `gen2` will be preferred.

> Other edge cases will be deterministic solved as well to ensure a consistent behavior, see the matching algorithm section.

Reviewer 3: ```suggestion
Other edge cases will be deterministic solved as well to ensure a consistent behavior. 
For further information, refer to the [matching algorithm](#matching-algorithm) section.
```

# CloudProfile Example
spec:
  capabilities:
    - hypervisorType: ["gen2", "gen1"]
    - network: ["accelerated", "standard"]
    - storageAccess: ["NVMe", "SCSI"]
    - secureBoot: ["secure", "none"]
    - bootMode: ["uefi-preferred", "uefi", "legacy-bios"]
    ...
```

Please note the following characteristics:
- The array of capabilities is ordered.
- The array of each capability is also ordered.
- The order defines the priority of the values in case of multiple supported images.

> - The first value in the array is the most preferred value.
> - The last value is the least preferred value.

Reviewer: ```suggestion
- The order of capabilities define their priority, .e.g, prefer `hypervisorType` over `network`.
- The order of capability values also define their priority, .e.g, prefer `gen2` over `gen1`.
```

# CloudProfile Example
spec:
  capabilities:

>     - name: hypervisorType

Reviewer 2: why is `architecture`, as a special kind of capability, not part of this example?

Author: it is so special, i missed it 👀

      values: ["gen2", "gen1"]
    - name: network
      values: ["accelerated", "standard"]
    - name: storageAccess
      values: ["NVMe", "SCSI"]
    - name: secureBoot
      values: ["secure", "none"]
    - name: bootMode
      values: ["uefi-preferred", "uefi", "legacy-bios"]
    ...
```

Please note the following characteristics:

> - The order of capabilities define their priority, .e.g, prefer `hypervisorType` over `network`.
> - The order of capability values also define their priority, .e.g, prefer `gen2` over `gen1`.

Reviewer: ```suggestion
- The order of capabilities define their priority, .e.g., prefer `hypervisorType` over `network`.
- The order of capability values also define their priority, .e.g., prefer `gen2` over `gen1`.
```

Typo :)

> - The order of capabilities define their priority, .e.g., prefer `hypervisorType` over `network`.
> - The order of capability values also define their priority, .e.g., prefer `gen2` over `gen1`.

Reviewer 2: ```suggestion
- The order of capabilities define their priority, e.g., prefer `hypervisorType` over `network`.
- The order of capability values also define their priority, e.g., prefer `gen2` over `gen1`.
```

return true
```

If multiple machine image versions are valid for a machine type, the selection is made based on the following rules:

> 1. The global capabilities array is evaluated in order. The first difference in the most preferred value of the capabilities determines the selection.

Reviewer: ```suggestion
1. The order of global capabilities are taken into consideration for choosing the best candidate. The first difference in the most preferred value of the capabilities determines the selection.
```
